Resize coordinates array in EdgesGeometry

dubejf opened the pull request:
The current position array is too large. It contains 6 extra `0` elements per removed coplanar edge.

dubejf committed:
Resize coordinates array in EdgesGeometry

The current position array is too large. It contains 6 extra `0` elements per removed coplanar edge.

gero3 commented:
Wouldn't it be better not make execute [LIne 73](https://github.com/dubejf/three.js/blob/patch-1/src/extras/geometries/EdgesGeometry.js#L73) before [line 57](https://github.com/dubejf/three.js/blob/patch-1/src/extras/geometries/EdgesGeometry.js#L57). And only increase numEdges then??

dubejf committed:
Track of exact number of unique edges in EdgesGeometry

dubejf commented:
I updated the patch as suggested.

WestLangley commented:
I liked what you had before -- a one-line addition. It might have not been the most efficient, but it was certainly clear.

What you have now is too confusing, IMHO.

@gero3 What do you think?

dubejf committed:
Simplify edge removal heuristic

Minor weak for non-manifold scenario.

Anytime an edge is non-manifold (more than two faces sharing an edge, e.g. a 'T'), always keep the edge in the EdgesGeometry.

With the previous commit, when a third co-planar face was found, the edge was marked co-planar. I'm not sure which is right but this algorithm is a bit simpler.

dubejf commented:
I agree. I've simplified the logic a bit.

dubejf commented:
About `h.coplanar !== undefined`:

The rule is: anytime an edge is non-manifold (more than two faces sharing an edge, e.g. a 'T'), always keep the edge in the EdgesGeometry.

dubejf committed:
Simplify expression

gero3 commented:
Isn't this even better??

https://gist.github.com/gero3/78658717eacbc04ffb16

benaadams commented:
@gero3 results.length * 2 * 3 rather than numEdges * 2 * 3?

dubejf commented:
@gero3 The gist would only work for edges with exactly two adjacent faces. You are not dealing with boundary edges, those with `face2 === undefined`. All those edges need to be included.

Also, you need to consider that an edge can be shared by more than 2 faces (forming a `T` or `+` pattern). A configuration that initially appears flat can become non-coplanar when a third face is added. You also have to be careful not including the same edge multiple times.

I don't think it that can be done efficiently with an array.

gero3 commented:
We need examples to test this for the special cases. That would make it easier verify the code.

dubejf commented:
Should I start a unit test suite for geometry processing?

WestLangley commented:
> I agree. I've simplified the logic a bit.

Like I said, I liked what you had before -- a one-line addition.

> Should I start a unit test suite for geometry processing?

That would be excellent.

    regular Geometry
    non-indexed BufferGeometry
    indexed BufferGeometry without drawcalls
    indexed BufferGeometry with drawcalls

That would be a good start...

mrdoob commented:
> > I agree. I've simplified the logic a bit.

> Like I said, I liked what you had before -- a one-line addition.

I agree with @WestLangley. Simple and readable.

gero3 commented:
There is a problem with the old code that makes me want the new code and the tests. When 2 faces are coplanar `(a & b)` and there is a third face `(c)` with the same edge which isn't coplanar, then the order matters because `a,b,c` doesn't give the same result as `a,c,b`.

mrdoob commented:
Hmm... The problem is that the code is a bit unreadable now :(

WestLangley commented:
`EdgesHelper` was not designed to support arbitrary meshes with three or more faces sharing a single edge.

If you want to support those cases, then you have to decide what the policy will be when some of the adjoining faces are coplanar and some are not.

gero3 commented:
Then, I fold. And we should use the one liner from the first commit.

dubejf committed:
Add unit tests

dubejf committed:
Revert EdgesGeometry to the initial patch. Modify unit test that is now failing.

dubejf commented:
Back to the simple one liner. Added unit tests.

dubejf committed:
Fix formating

benaadams commented on src/extras/geometries/EdgesGeometry.js:
```diff
@@ -86,6 +86,8 @@ THREE.EdgesGeometry = function ( geometry, thresholdAngle ) {
 
 	}
 
+	coords = coords.subarray( 0, index );
```
Should be: 
```
coords = new Float32Array( coords.subarray( 0, index ) );
```
Else the underlying ```ArrayBuffer``` behind the coords will still be the full allocation size.

benaadams replied:
Could also just do something simpler and drop counting and array resizing:
````javascript
	var vertices = geometry2.vertices;
	var faces = geometry2.faces;

	for ( var i = 0, l = faces.length; i < l; i ++ ) {

		var face = faces[ i ];

		for ( var j = 0; j < 3; j ++ ) {

			edge[ 0 ] = face[ keys[ j ] ];
			edge[ 1 ] = face[ keys[ ( j + 1 ) % 3 ] ];
			edge.sort( sortFunction );

			var key = edge.toString();

			if ( hash[ key ] === undefined ) {

				hash[ key ] = { vert1: edge[ 0 ], vert2: edge[ 1 ], face1: i, face2: undefined };

			} else {

				hash[ key ].face2 = i;

			}

		}

	}

	var coords = [];

	for ( var key in hash ) {

		var h = hash[ key ];

		if ( h.face2 === undefined || faces[ h.face1 ].normal.dot( faces[ h.face2 ].normal ) <= thresholdDot ) {

			var vertex = vertices[ h.vert1 ];
			coords.push( vertex.x );
			coords.push( vertex.y );
			coords.push( vertex.z );

			vertex = vertices[ h.vert2 ];
			coords.push( vertex.x );
			coords.push( vertex.y );
			coords.push( vertex.z );

		}

	}

	coords = new Float32Array( coords );
````

mrdoob replied:
@WestLangley thoughts?

WestLangley replied:
Yes, that looks even better.

dubejf replied:
@WestLangley Sorry, which version are you talking about? The cloned Float32Array or the temporary JS array?

WestLangley replied:
I have not tested it, but I was refering to @benaadams's suggestion to "do something simpler". What do you think?

dubejf replied:
Okay - that's the second of @benaadams's two suggestions. I wanted to avoid over-allocating and extra copying in my initial attempts, but, I don't think it matters too much here. I'll update the patch.

benaadams replied:
>  I wanted to avoid over-allocating and extra copying

Won't really matter as lots of temporary alloc already already going on there, so it will all get GC'd together; also assume it will be used in setup so wouldn't be adding to the performance critical per frame memory pressure?

dubejf committed:
Simplify tracking of EdgesGeometry coordinates

dubejf commented:
@mrdoob Done. Could you please run the unit tests (`unittests_three.html`) after merging to make sure everything is still okay.

mrdoob commented:
@dubejf will do! many thanks!

mrdoob commented:
Hmm... Maybe is an issue with the current state of the `dev` branch. But all the unitests fail...

CodingFabian commented:
@mrdoob for me only EdgesGeometry tests fail. 42

mrdoob commented:
Ah yes, sorry. I meant all the EdgesGeometry test fail.
